File: cpp/se/AeDomainTransferRegistrantCommand.cpp
#include "se/AeDomainTransferRegistrantCommand.hpp"
#include "se/XMLGregorianCalendar.hpp"

#include "se/AeExtension.hpp"
#include "se/AeDomainObjectType.hpp"
#include "se/RegistrantTransferCommandType.hpp"
#include "se/CommandType.hpp"
#include "se/Period.hpp"
#include "common/ErrorPkg.hpp"

#include "se/EPPDateFormatter.hpp"

#include "xml/XMLHelper.hpp"

namespace {
    Extension& aeExtension() {
        static Extension* aeExt = new AeExtension();
        return *aeExt;
    }

    const RegistrantTransferCommandType rtrnType;
    const AeDomainObjectType aedomType;
} // anonymous namespace

AeDomainTransferRegistrantCommand::AeDomainTransferRegistrantCommand (
        const std::string& name,
        const XMLGregorianCalendar& curExpDate,
        const std::string& eligibilityType,
        int policyReason,
        const std::string& registrantName,
        const std::string& explanation,
        const std::string* registrantID,
        const std::string* registrantIDType,
        const std::string* eligibilityName,
        const std::string* eligibilityID,
        const std::string* eligibilityIDType,
        const Period* period) : ProtocolExtensionCommand(
            &rtrnType, &aedomType, name, aeExtension())
{
    if ((registrantID && registrantIDType == NULL)
            || (registrantIDType == NULL && registrantIDType)
            || (eligibilityName && (eligibilityID == NULL || eligibilityIDType == NULL))
            || (eligibilityName == NULL && (eligibilityID || eligibilityIDType)))
    {
        // If provided, a registrantID must have a type.
        // If provided, an eligibilityName must have both an eligibilityID and type.
        throw IllegalArgException(
                ErrorPkg::getMessage("se.domain.registrantTransfer.ae.missing_arg"));
    }

    DOMElement *element;

    std::string curExpDateStr = EPPDateFormatter::toXSDate(curExpDate);
    XMLHelper::setTextContent(
            xmlWriter->appendChild(objElement, "curExpDate"), curExpDateStr);

    if (period)
        period->appendPeriod(xmlWriter, objElement);

    DOMElement *aeProperties = xmlWriter->appendChild(objElement, "aeProperties");

    XMLHelper::setTextContent(
            xmlWriter->appendChild(aeProperties, "registrantName"), registrantName);

    if (registrantID)
    {
        element = xmlWriter->appendChild(aeProperties, "registrantID");
        XMLHelper::setTextContent(element, *registrantID);
        XMLHelper::setAttribute(element, "type", *registrantIDType);
    }
    XMLHelper::setTextContent
        (xmlWriter->appendChild(aeProperties, "eligibilityType"), eligibilityType);

    if (eligibilityName)
    {
        XMLHelper::setTextContent(
                xmlWriter->appendChild(aeProperties, "eligibilityName"), *eligibilityName);

        element = xmlWriter->appendChild(aeProperties, "eligibilityID");
        XMLHelper::setTextContent(element, *eligibilityID);
        XMLHelper::setAttribute(element, "type", *eligibilityIDType);
    }

    XMLHelper::setTextContent(
            xmlWriter->appendChild(aeProperties, "policyReason"),
            policyReason);

    XMLHelper::setTextContent(
            xmlWriter->appendChild(objElement, "explanation"),
            explanation);
}

File: src/main/java/com/ausregistry/jtoolkit2/se/AeDomainTransferRegistrantCommand.java
package com.ausregistry.jtoolkit2.se;

import java.util.GregorianCalendar;

import com.ausregistry.jtoolkit2.EPPDateFormatter;
import com.ausregistry.jtoolkit2.ErrorPkg;
import org.w3c.dom.Element;

/**
 * In cases where the legal registrant of a .ae domain name has changed, this
 * class should be used to request a transfer of registrant. This is a different
 * action to correcting extension data which was originally specified
 * incorrectly, and should only be used in the situation described. This command
 * will result in the validity period of the domain name being updated and the
 * requesting client being charged the usual create fee upon success of this
 * operation. Use this class to generate a standards-compliant XML document,
 * given simple input parameters. The toXML method in Command serialises this
 * object to XML.
 *
 */
public final class AeDomainTransferRegistrantCommand extends ProtocolExtensionCommand {

    private static final long serialVersionUID = -7263014552265126595L;
    private static final CommandType RETURN_TYPE = new AeRegistrantTransferCommandType();

    public AeDomainTransferRegistrantCommand(String name, GregorianCalendar curExpDate,
            String registrantName, String eligibilityType, int policyReason, String explanation) {

        this(name, curExpDate, null, eligibilityType, policyReason, registrantName, null, null,
                null, null, null, explanation);
    }

    /**
     * Request that the named .ae domain name be transferred to the legal entity
     * specified by the given ae extension data.
     *
     * @param name
     *            The domain name to transfer. Required.
     *
     * @param curExpDate
     *            The current expiry of the identified domain name. This is
     *            required in order to prevent repeated transfer of the name due
     *            to protocol transmission failures. Required.
     *
     * @param period
     *            The desired new validity period, starting from the time the
     *            transfer completes successfully. Optional.
     * @param eligibilityType Required.
     * @param policyReason
     * @param registrantName Required.
     * @param registrantID
     * @param registrantIDType
     * @param eligibilityName
     * @param eligibilityID
     * @param eligibilityIDType
     *
     * @param explanation
     *            An explanation of how the transfer was effected.
     * @throws IllegalArgumentException if {@code curExpDate}, {@code eligibilityType} or {@code registrantIDType}
     * is {@code null}.
     */
    public AeDomainTransferRegistrantCommand(String name, GregorianCalendar curExpDate,
            Period period, String eligibilityType, int policyReason, String registrantName,
            String registrantID, String registrantIDType, String eligibilityName,
            String eligibilityID, String eligibilityIDType, String explanation) {

        super(RETURN_TYPE, ExtendedObjectType.AE_DOMAIN, name);

        if (curExpDate == null || eligibilityType == null || registrantName == null) {
            throw new IllegalArgumentException(
                    ErrorPkg.getMessage("se.domain.registrantTransfer.ae.missing_arg"));
        }

        // A Registrant ID, if present, must be qualified by a type.
        assert (registrantID == null && registrantIDType == null)
                || (registrantID != null && registrantIDType != null);

        if (eligibilityName != null) {
            assert eligibilityID != null;
            assert eligibilityIDType != null;
        }

        Element element;

        String curExpDateStr = EPPDateFormatter.toXSDate(curExpDate);
        xmlWriter.appendChild(objElement, "curExpDate").setTextContent(curExpDateStr);

        if (period != null) {
            period.appendPeriod(xmlWriter, objElement);
        }

        Element aeProperties = xmlWriter.appendChild(objElement, "aeProperties");

        xmlWriter.appendChild(aeProperties, "registrantName").setTextContent(registrantName);
        if (registrantID != null) {
            element = xmlWriter.appendChild(aeProperties, "registrantID");
            element.setTextContent(registrantID);
            element.setAttribute("type", registrantIDType);
        }
        xmlWriter.appendChild(aeProperties, "eligibilityType").setTextContent(eligibilityType);
        if (eligibilityName != null) {
            // Having an eligibility name implies that the entity must also
            // have an identifier for their eligibility name, specified by an
            // eligibility ID with required type attribute.
            xmlWriter.appendChild(aeProperties, "eligibilityName").setTextContent(eligibilityName);
            element = xmlWriter.appendChild(aeProperties, "eligibilityID");
            element.setTextContent(eligibilityID);
            element.setAttribute("type", eligibilityIDType);
        }
        String policyReasonStr = String.valueOf(policyReason);
        xmlWriter.appendChild(aeProperties, "policyReason").setTextContent(policyReasonStr);
        xmlWriter.appendChild(objElement, "explanation").setTextContent(explanation);
    }

    @Override
    protected Extension getExtension() {
        return ExtensionImpl.AE;
    }
}
